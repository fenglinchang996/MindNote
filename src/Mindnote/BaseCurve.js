import React from "react";

const BaseCurve = (props) => {
  const { curveData } = props;
  const { id, start, end, startControl, endControl, style } = curveData;
  // Path info (d)
  const pathInfo = `M ${start.x} ${start.y} C ${startControl.x} ${startControl.y}, ${endControl.x} ${endControl.y}, ${end.x} ${end.y}`;

  return (
    <g>
      <ArrowMarker />
      <path d={pathInfo} style={style} markerEnd="url(#Arrow)" />
    </g>
  );
};

const ArrowMarker = (props) => {
  return (
    <defs>
      <marker
        id="Arrow"
        markerWidth="3"
        markerHeight="3"
        refX="1.5"
        refY="1.5"
        orient="auto"
      >
        <polygon points="0,0 3,1.5 0,3" fill="blue" />
      </marker>
    </defs>
  );
};

export default BaseCurve;
