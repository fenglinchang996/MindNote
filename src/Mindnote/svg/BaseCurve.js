import React from "react";

const BaseCurve = (props) => {
  const { curveData } = props;
  const { id, start, end, startControl, endControl, style } = curveData;
  // Path info (d)
  const pathInfo = `M ${start.x} ${start.y} C ${startControl.x} ${startControl.y}, ${endControl.x} ${endControl.y}, ${end.x} ${end.y}`;

  return (
    <g>
      <ThickArrowMarker id={id} fill={style.stroke} />
      <ArrowMarker id={id} stroke={style.stroke} />
      <path d={pathInfo} style={style} markerEnd={`url(#ThickArrow-${id})`} />
    </g>
  );
};

const ThickArrowMarker = (props) => {
  const { id, fill } = props;
  return (
    <defs>
      <marker
        id={`ThickArrow-${id}`}
        markerWidth="3"
        markerHeight="3"
        refX="1.5"
        refY="1.5"
        orient="auto"
      >
        <polygon points="0,0 3,1.5 0,3" fill={fill} />
      </marker>
    </defs>
  );
};

const ArrowMarker = (props) => {
  const { id, stroke } = props;
  return (
    <defs>
      <marker
        id={`Arrow-${id}`}
        markerWidth="6"
        markerHeight="6"
        refX="4"
        refY="3"
        orient="auto"
      >
        <polyline
          points="1,1 5,3, 1,5"
          strokeWidth={1}
          strokeLinecap="round"
          stroke={stroke}
          fill="none"
        />
      </marker>
    </defs>
  );
};

export default BaseCurve;
