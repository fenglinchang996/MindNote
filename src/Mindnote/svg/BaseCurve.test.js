import React from "react";
import renderer from "react-test-renderer";
import BaseCurve from "./BaseCurve";

test("BaseCurve", () => {
  const curveData = {
    id: "test",
    start: { x: 480, y: 295 },
    startControl: { x: 479, y: 331 },
    endControl: { x: 478, y: 350 },
    end: { x: 478, y: 386 },
    level: 1,
    style: {
      fill: "none",
      stroke: "blue",
      strokeWidth: 5,
      strokeLinecap: "round",
    },
  };
  const component = renderer.create(<BaseCurve curveData={curveData} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
