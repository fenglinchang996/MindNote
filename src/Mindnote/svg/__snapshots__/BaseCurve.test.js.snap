// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseCurve 1`] = `
<g>
  <defs>
    <marker
      id="Arrow-test"
      markerHeight="3"
      markerWidth="3"
      orient="auto"
      refX="1.5"
      refY="1.5"
    >
      <polygon
        fill="blue"
        points="0,0 3,1.5 0,3"
      />
    </marker>
  </defs>
  <path
    d="M 480 295 C 479 331, 478 350, 478 386"
    markerEnd="url(#Arrow-test)"
    style={
      Object {
        "fill": "none",
        "stroke": "blue",
        "strokeLinecap": "round",
        "strokeWidth": 5,
      }
    }
  />
</g>
`;
